File: src/risk_management/position_sizer.py
"""Position sizing utilities."""

from typing import Dict, Any
from loguru import logger


class PositionSizer:
    """
    Calculate optimal position sizes based on various methodologies.
    """

    def __init__(self, config: Dict[str, Any]):
        """
        Initialize position sizer.

        Args:
            config: Configuration dictionary
        """
        self.config = config
        self.risk_config = config.get('risk', {})

    def fixed_percentage(
        self,
        capital: float,
        percentage: float
    ) -> float:
        """
        Calculate position size as a fixed percentage of capital.

        Args:
            capital: Available capital
            percentage: Percentage of capital to allocate (0-100)

        Returns:
            Position value
        """
        return capital * (percentage / 100)

    def kelly_criterion(
        self,
        capital: float,
        win_rate: float,
        avg_win: float,
        avg_loss: float
    ) -> float:
        """
        Calculate position size using Kelly Criterion.

        Args:
            capital: Available capital
            win_rate: Historical win rate (0-1)
            avg_win: Average winning trade amount
            avg_loss: Average losing trade amount

        Returns:
            Position value
        """
        if avg_loss == 0:
            return 0

        win_loss_ratio = avg_win / abs(avg_loss)
        kelly_pct = (win_rate * win_loss_ratio - (1 - win_rate)) / win_loss_ratio

        # Use fractional Kelly (typically 25-50% of full Kelly for safety)
        kelly_pct = kelly_pct * 0.25

        # Ensure kelly_pct is between 0 and 1
        kelly_pct = max(0, min(1, kelly_pct))

        return capital * kelly_pct

    def risk_based(
        self,
        capital: float,
        risk_per_trade: float,
        entry_price: float,
        stop_loss_price: float
    ) -> float:
        """
        Calculate position size based on risk amount and stop loss.

        Args:
            capital: Available capital
            risk_per_trade: Maximum risk per trade as percentage (0-100)
            entry_price: Entry price
            stop_loss_price: Stop loss price

        Returns:
            Position size in units
        """
        risk_amount = capital * (risk_per_trade / 100)
        risk_per_unit = abs(entry_price - stop_loss_price)

        if risk_per_unit == 0:
            return 0

        position_size = risk_amount / risk_per_unit
        return position_size

    def volatility_based(
        self,
        capital: float,
        target_risk: float,
        volatility: float
    ) -> float:
        """
        Calculate position size based on volatility (ATR).

        Args:
            capital: Available capital
            target_risk: Target risk as percentage (0-100)
            volatility: Asset volatility (e.g., ATR)

        Returns:
            Position value
        """
        if volatility == 0:
            return 0

        position_value = (capital * target_risk / 100) / volatility
        return position_value

File: rust/src/risk_management/position_sizer.rs
//! Position sizing utilities

/// Position sizer for calculating optimal position sizes
pub struct PositionSizer;

impl PositionSizer {
    /// Calculate position size as fixed percentage of capital
    pub fn fixed_percentage(capital: f64, percentage: f64) -> f64 {
        capital * (percentage / 100.0)
    }

    /// Calculate position size using Kelly Criterion
    pub fn kelly_criterion(
        capital: f64,
        win_rate: f64,
        avg_win: f64,
        avg_loss: f64,
    ) -> f64 {
        if avg_loss == 0.0 {
            return 0.0;
        }

        let win_loss_ratio = avg_win / avg_loss.abs();
        let kelly_pct = (win_rate * win_loss_ratio - (1.0 - win_rate)) / win_loss_ratio;

        // Use fractional Kelly (25% of full Kelly for safety)
        let kelly_pct = (kelly_pct * 0.25).clamp(0.0, 1.0);

        capital * kelly_pct
    }

    /// Calculate position size based on risk and stop loss
    pub fn risk_based(
        capital: f64,
        risk_per_trade: f64,
        entry_price: f64,
        stop_loss_price: f64,
    ) -> f64 {
        let risk_amount = capital * (risk_per_trade / 100.0);
        let risk_per_unit = (entry_price - stop_loss_price).abs();

        if risk_per_unit == 0.0 {
            return 0.0;
        }

        risk_amount / risk_per_unit
    }

    /// Calculate position size based on volatility
    pub fn volatility_based(capital: f64, target_risk: f64, volatility: f64) -> f64 {
        if volatility == 0.0 {
            return 0.0;
        }

        (capital * target_risk / 100.0) / volatility
    }
}
